### ISSUE TITLE
Add login retries

### ISSUE THREAD

**Issue by Sing-Li** [RocketChat/hubot-rocketchat#32]
Current login try once and fails, making login problem difficult to debug.

Include more extensive logging. Help to identify as much as possible, the cause of the failure.

Include details (if available) of networking stack failure - for example, invalid SSL certs.

**Comment by timkinnane** [comment]
The new SDK driver's `connect` method now accepts an options object, with a `timeout` value.

https://github.com/RocketChat/Rocket.Chat.js.SDK/blob/c8c48425f30df99741b54fecffb1718e16e1650f/src/lib/driver.ts#L40

It is not used in the adapter currently, but since it's now async, we could possibly add a connection loop, with a set number of attempts by awaiting the connection with a shorter timeout, exiting on successful promise resolve or on reaching the maximum attempts.

However, I don't know why a second attempt would work where the first failed with the same values, unless the server was reset and comes back between attempts. Even then, I don't know if maybe one long attempt would do just as well as two short attempts.

Another reason I didn't implement timeout yet, is because the current version of Asteroid cannot actually cancel connections, the timeout just abandons the background process, but I believe it still runs and could hypothetically end up getting a connection. So by setting a low timeout and multiple tries, you might end up creating multiple connections, but only using one of them. I have no idea the implications, just recording this for future debuggers.

In summary, I don't know if it's there's much value at this point. Maybe when we [upgrade the DDP module](https://github.com/RocketChat/Rocket.Chat.js.SDK/issues/4).

**Comment by timkinnane** [comment]
I'm going to close this now with the link through to the SDK issue to hopefully enhance this in future with a new DDP module.

See https://github.com/RocketChat/Rocket.Chat.js.SDK/issues/4
